File: 0041-first-missing-positive/0041-first-missing-positive.java
class Solution {
    public int firstMissingPositive(int[] nums) {
        Set<Integer> set = new HashSet<>();
        for(int num : nums) {
            if(num > 0 && num <= nums.length) {//vvimp
                set.add(num);
            }
        }
        for(int i=1; i<=nums.length; i++) {//vvimp
            if(!set.contains(i)) {
                return i;
            }
        }
        return nums.length +1;
    }
}

File: 0041-first-missing-positive/0041-first-missing-positive.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        nums.append(0)
        n = len(nums)
        for i in range(len(nums)):  # delete those useless elements
            if nums[i] < 0 or nums[i] >= n:
                nums[i] = 0

        for i in range(len(nums)):  # use the index as the hash to record the frequency of each number
            nums[nums[i] % n] += n

        for i in range(1, len(nums)):
            if nums[i] // n == 0:
                return i

        return n
